test/jasmine/tests/scatter_test.js

```diff
@@ -629,6 +630,98 @@ describe('end-to-end scatter tests', function() {
         .then(done);
     });
 
+    it('should keep layering correct as mode & fill change', function(done) {
+        var fillCase = {name: 'fill', edit: {mode: 'none', fill: 'tonexty'}};
+        var i, j;
+
+        var cases = [fillCase];
+        var modeParts = ['lines', 'markers', 'text'];
+        for(i = 0; i < modeParts.length; i++) {
+            var modePart = modeParts[i];
+            var prevCasesLength = cases.length;
+
+            cases.push({name: modePart, edit: {mode: modePart, fill: 'none'}});
+            for(j = 0; j < prevCasesLength; j++) {
+                var prevCase = cases[j];
+                cases.push({
+                    name: prevCase.name + '_' + modePart,
+                    edit: {
+                        mode: (prevCase.edit.mode === 'none' ? '' : (prevCase.edit.mode + '+')) + modePart,
+                        fill: prevCase.edit.fill
+                    }
+                });
+            }
+        }
+
+        // visit each case N times, in an order that covers each *transition*
+        // from any case to any other case.
+        var indices = [];
+        var curIndex = 0;
+        for(i = 1; i < cases.length; i++) {
+            for(j = 0; j < cases.length; j++) {
+                indices.push(curIndex);
+                curIndex = (curIndex + i) % cases.length;
+            }
+        }
```

Comment:
Notice here we have 15 cases, so we're covering 15*14 = 210 transitions between cases. Takes about 4 seconds locally but totally worth it, as it uncovered a few other layering problems besides the lines/markers one that prompted this effort.

Reply:
Oh hmm, actually this would work perfectly if `cases.length` were prime, but as it is I duplicated some and missed some others. I wonder if there is a generic efficient way to do this with a composite number of cases? Anyway *most* of the transitions are covered so I think this is fine as is but I am curious now...

Reply:
Amazing testing!

Reply:
Found an algorithm that works efficiently for any `cases.length` -> https://github.com/plotly/plotly.js/pull/2978/commits/69888187c9de68a3d8a6d75cd1bb84e6c0c66f1a